File: solve_wordle.py
import string
import pickle
import sys

from itertools import product
from scipy.stats import entropy
from os.path import exists
from functools import partial

def load_words():
    # with open('words_alpha.txt') as word_file:
    with open('sowpods.txt') as word_file:
        valid_words = list(word_file.read().split())

    return valid_words

def has_letter_at_pos(word, letter, pos):
    return word[pos] == letter

def has_letter_not_at_pos(word, letter, pos):
    if has_letter_at_pos(word, letter, pos):
        return False
    return word.find(letter) != -1

def does_not_have_letter(word, letter, pos):
    return word.find(letter) == -1

def calc_entropy_for_word(query, all_words):
    # what are all the ways things could turn out?
    # - 243 combinations
    # - 12.5k words
    # - 5 checks
    # - thats ~ 15 million checks per word...
    counts = []
    for comb in product([has_letter_at_pos, has_letter_not_at_pos, does_not_have_letter], repeat=5):
        count = 0
        for word in all_words:
            if all([comb[i](word, query[i], i) for i in range(5)]):
                # print("found one!" + word)
                count += 1
        counts.append(count)
    probs = [float(count) / len(all_words) for count in counts]
    # print(probs[0:10])
    return entropy(probs, base=2)

def parse_constraints_string(constraints_string):
    if len(constraints_string) == 0:
        return []
    constraints = []
    constraint_order = [has_letter_at_pos, has_letter_not_at_pos, does_not_have_letter]
    for i in range(0, len(constraints_string), 2):
        letter = constraints_string[i]
        idx = int(constraints_string[i+1])-1
        constraints.append(partial(constraint_order[idx],letter=letter, pos=i/2))
    return constraints

def get_best_word(word_list, use_cache):
    entrop_dict = {}
    if use_cache and exists('entrop.pickle'):
        with open('entrop.pickle', 'rb') as handle:
            entrop_dict = pickle.load(handle)
            print("loading entropy from pickle with length: ", len(entrop_dict))


    count = len(entrop_dict) # use this for checkpointing
    for query in word_list:
        if entrop_dict.has_key(query):
            # print("skipping query that's already computed: ", query)
            continue
        entrop_dict[query] = calc_entropy_for_word(query, word_list)
        if use_cache and count % 20 == 0:
            print(query, " has entropy: ", entrop_dict[query])
            print("checkpointing entropy at count: ", count)
            with open('entrop.pickle', 'wb') as handle:
                pickle.dump(entrop_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)
        count += 1

    best_word = max(entrop_dict, key=entrop_dict.get)
    entrop = entrop_dict[best_word]
    return (best_word, entrop)

if __name__ == '__main__':
    english_words = load_words()

    # count number of 5-letter words
    five_letter_words = [word for word in english_words if len(word) == 5]
    print(len(five_letter_words))

    first_guess, entrop = get_best_word(five_letter_words, use_cache=True)
    print("let's guess: ", first_guess, " which has entropy: ", entrop)

    constrained_words = five_letter_words
    constraints = []
    while True:
        constraints_string = raw_input("Please input constraint string: (ex. t1e2a2r3s3 would mean the word contains a 't' in the correct position, 'e' and 'a' in wrong positions and does not contain 'r' or 's')\n")
        # todo: validation checking
        new_constraints = parse_constraints_string(constraints_string)
        constraints.extend(new_constraints)
        constrained_words = [word for word in constrained_words if all([constraint(word) for constraint in constraints])]

        if len(constrained_words) == 0:
            print("no words found matching all constraints :(")
            sys.exit()
        elif len(constrained_words) == 1:
            print("found only one choice: ", constrained_words[0])
            sys.exit()

        next_guess, entrop = get_best_word(constrained_words, use_cache=False)
        print("let's guess: ", next_guess, "which has entropy: ", entrop, " while max entropy for this list of words is ", entropy([1.0/len(constrained_words) for i in range(len(constrained_words))], base=2))

File: cpp/solve_wordle.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <math.h>
#include <numeric>

#include "cppitertools/product.hpp"
#include "utils.h"

std::vector<std::function<bool(const std::string&)>> parse_constraints_string(const std::string& constraints_string) {
  if (constraints_string.empty()) {
    return {};
  }
  std::vector<std::function<bool(const std::string&)>> constraints = {};

  std::vector<std::function<bool(const std::string&, const char, int)>> raw_constraints;

  raw_constraints.push_back(has_letter_at_pos);
  raw_constraints.push_back(has_letter_not_at_pos);
  raw_constraints.push_back(does_not_have_letter);

  for (int i = 0; i < constraints_string.size(); i+= 2) {
    char letter = constraints_string.at(i);
    int idx = (int)(constraints_string.at(i+1))-(int)('0') - 1;
    constraints.push_back(std::bind(raw_constraints.at(idx), std::placeholders::_1, letter, i/2));
  }
  return constraints;
}

int main() {
  // get list of words
  std::vector<std::string> guess_words = load_guess_words();
  std::vector<std::string> sol_words = load_sol_words();
  std::cout << "number of five letter guess words: " << guess_words.size() << std::endl;

  std::vector<int> constrained_guess_idxs(guess_words.size());
  std::iota(std::begin(constrained_guess_idxs), std::end(constrained_guess_idxs), 0);

  std::vector<int> constrained_sol_idxs(sol_words.size());
  std::iota(std::begin(constrained_sol_idxs), std::end(constrained_sol_idxs), 0);


  auto [guess,ent] = get_best_word(guess_words, constrained_guess_idxs, sol_words, constrained_sol_idxs, /*use_cache=*/true);
  std::cout << guess << " has highest entropy of " << ent << std::endl;

  std::string constraints_string;
  std::vector<std::function<bool(const std::string&)>> constraints;
  while (true) {
    std::cout << "Please input constraint string: (ex. t1e2a2r3s3 would mean the word contains a 't' in the correct position, 'e' and 'a' in wrong positions and does not contain 'r' or 's')\n" << std::endl;
    std::cin >> constraints_string;
    std::vector<std::function<bool(const std::string&)>> new_constraints = parse_constraints_string(constraints_string);
    constraints.insert(constraints.end(), new_constraints.begin(), new_constraints.end());
    // recompute constrained_word_idxs
    {
      std::vector<int> temp;
      for (const int idx : constrained_guess_idxs) {
	bool failed_constraint = false;
	const auto& guess = guess_words.at(idx);
	for (const auto& constraint : constraints) {
	  if (!constraint(guess)) {
	    failed_constraint = true;
	    break;
	  }
	}
	if (!failed_constraint) {
	  temp.push_back(idx);
	}
      }
      constrained_guess_idxs = temp;
    }

    // do the same for sols
    {
      std::vector<int> temp;
      for (const int idx : constrained_sol_idxs) {
	bool failed_constraint = false;
	const auto& sol = sol_words.at(idx);
	for (const auto& constraint : constraints) {
	  if (!constraint(sol)) {
	    failed_constraint = true;
	    break;
	  }
	}
	if (!failed_constraint) {
	  temp.push_back(idx);
	}
      }
      constrained_sol_idxs = temp;
    }

    if (constrained_sol_idxs.size() == 1) {
      std::cout << "found only one choice: " << sol_words.at(constrained_sol_idxs.at(0)) << std::endl;
      return 0;
    } else if (constrained_sol_idxs.size() == 0) {
      std::cout << "no words found matching all constraints. either a bug or vocab isn't big enough" << std::endl;
      return 0;
    }
    auto [next_guess, ent] = get_best_word(guess_words, constrained_guess_idxs, sol_words, constrained_sol_idxs, /*use_cache=*/false);
    std::cout << "let's guess: " << next_guess << " which has entropy: " << ent << std::endl;
  }
}
